### PR TITLE
[core] Allow cache and refresh partitions for CachingCatalog

### LINKED ISSUE #4205: [Feature] In paimon catalog, add partition  query and cache
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.

### Motivation

In the OLAP scenario, it is necessary to obtain the maximum partition of the table.Currently, the partition is obtained through createReader, which takes about 500ms. Therefore, it is necessary to add partition query and cache in paimon catalog

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by mxdzs0612**
### Purpose

Linked issue: close #4205

### Tests

testPartitionCache

### API and Format

CachingCatalog.getPartitions
CachingCatalog.refreshPartitions

### Documentation

Allow cache and refresh partitions for CachingCatalog.

### CODE REVIEW

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Can you add an option? For example, it is called `cache.partition.max-num`, which defaults to 0, meaning no cache partition.

File: paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java [hunk 1]
```diff
@@ -227,6 +244,47 @@ private void putTableCache(Identifier identifier, Table table) {
...
+                && table instanceof FileStoreTable
+                && !table.partitionKeys().isEmpty();
+    }
+
+    @Override
+    public void dropPartition(Identifier identifier, Map<String, String> partitions)
+            throws TableNotExistException, PartitionNotExistException {
+        wrapped.dropPartition(identifier, partitions);
+        if (partitionCache != null) {
+            partitionCache.invalidate(identifier);
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java:284]: Why invalidate all partitions of this identifier in the cache?

File: paimon-core/src/test/java/org/apache/paimon/catalog/CachingCatalogTest.java [hunk 2]
```diff
@@ -209,17 +216,38 @@ public void testCacheExpirationEagerlyRemovesSysTables() throws Exception {
...
+                        singletonList("f0"),
+                        emptyList(),
+                        Collections.emptyMap(),
+                        "");
+        catalog.createTable(tableIdent, schema, false);
+        List<PartitionEntry> partitionEntryList = catalog.getPartitions(tableIdent);
+        assertThat(catalog.partitionCache().asMap().containsKey(tableIdent));
+        List<PartitionEntry> partitionEntryListFromCache =
+                catalog.partitionCache().getIfPresent(tableIdent);
+        assertThat(partitionEntryListFromCache.containsAll(partitionEntryList));
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/catalog/CachingCatalogTest.java:248]: partitionEntryListFromCache is nullable

File: paimon-core/src/test/java/org/apache/paimon/catalog/CachingCatalogTest.java [hunk 3]
```diff
@@ -209,17 +216,38 @@ public void testCacheExpirationEagerlyRemovesSysTables() throws Exception {
...
+        Schema schema =
+                new Schema(
+                        RowType.of(VarCharType.STRING_TYPE, VarCharType.STRING_TYPE).getFields(),
+                        singletonList("f0"),
+                        emptyList(),
+                        Collections.emptyMap(),
+                        "");
+        catalog.createTable(tableIdent, schema, false);
+        List<PartitionEntry> partitionEntryList = catalog.getPartitions(tableIdent);
+        assertThat(catalog.partitionCache().asMap().containsKey(tableIdent));
```
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/catalog/CachingCatalogTest.java:245]: Assertions.assertTrue

File: paimon-core/src/test/java/org/apache/paimon/catalog/CachingCatalogTest.java (hunk 2 above)
Comment (wwj6591812) [inline paimon-core/src/test/java/org/apache/paimon/catalog/CachingCatalogTest.java:248]: Assertions.assertTrue

File: paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java [hunk 4]
```diff
@@ -227,6 +244,47 @@ private void putTableCache(Identifier identifier, Table table) {
...
+            partitions = partitionCache.getIfPresent(identifier);
+            if (partitions == null || partitions.isEmpty()) {
+                partitions = this.refreshPartitions(identifier);
+            }
+            return partitions;
+        }
+        return ((FileStoreTable) table).newSnapshotReader().partitionEntries();
+    }
+
+    public List<PartitionEntry> refreshPartitions(Identifier identifier)
```
[the remark this answers is not in the document]
Comment (mxdzs0612) [reply paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java:260]: Other engine may call this method regularly, so it should be public.

File: paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java [hunk 5]
```diff
@@ -227,6 +244,47 @@ private void putTableCache(Identifier identifier, Table table) {
...
+            throws TableNotExistException {
+        Table table = this.getTable(identifier);
+        List<PartitionEntry> partitions =
+                ((FileStoreTable) table).newSnapshotReader().partitionEntries();
+        if (enablePartitionCache(table)
+                && partitionCache.asMap().values().stream().mapToInt(List::size).sum()
+                        < this.partitionMaxNum) {
+            partitionCache.put(identifier, partitions);
```
[the remark this answers is not in the document]
Comment (mxdzs0612) [reply paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java:268]: No more partitions will be put in util some are expired, I suppose it tolerable.

File: paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java (hunk 1 above)
Comment (mxdzs0612) [reply paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java:284]: Can't find a better way, any ideas?

File: paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java [hunk 6]
```diff
@@ -111,7 +119,15 @@ public CachingCatalog(
                         .expireAfterAccess(expirationInterval)
                         .ticker(ticker)
                         .build();
+        this.partitionCache =
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java:122]: if (cachedPartitionMaxNum == 0) partitionCache = null;

File: paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java [hunk 7]
```diff
@@ -111,7 +119,15 @@ public CachingCatalog(
...
+                Caffeine.newBuilder()
+                        .softValues()
+                        .executor(Runnable::run)
+                        .expireAfterAccess(expirationInterval)
+                        .ticker(ticker)
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java:127]: Can you use `maximumWeight(cachedPartitionMaxNum)`?
You can also set `weigher`.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**